web/shared/src/typeahead.ts

```diff
@@ -73,6 +105,57 @@ export function query_matches_string(
     return source_str.startsWith(query) || source_str.includes(split_char + query);
 }
 
+// This function attempts to match a query in any order with a source text.
+// * query is the user-entered search query
+// * source_str is the string we're matching in, e.g. a user's name
+// * split_char is the separator for this syntax (e.g. ' ').
+export function query_matches_string_in_any_order(
+    query: string,
+    source_str: string,
+    split_char: string,
+): boolean {
+    source_str = source_str.toLowerCase();
+    source_str = remove_diacritics(source_str);
+
+    query = query.toLowerCase();
+    query = remove_diacritics(query);
+
+    // We sort words in the query and source string then try to find the
+    // best match for each query word, using `upper_bound_lexicographically`.
+    // As soon as we find a query word with no unique match, we return false,
+    // else return true at the end.
+    const search_words = query.split(split_char).filter(Boolean);
+    const source_words = source_str.split(split_char).filter(Boolean);
+    if (search_words.length > source_words.length) {
+        // If there are more words in the query than in the source string
+        // we immediately return false.
+        return false;
+    }
+    // Search words are sorted in reverse lexicographical order while source
+    // words are ordered lexicographically.
+    search_words.sort().reverse();
+    source_words.sort();
+    for (const word of search_words) {
+        // `match_index` is the index of the best possible match of `word`.
+        const match_index = upper_bound_lexicographically(word, source_words) - 1;
+        if (match_index === -1) {
+            // We return false if no match was found for `word`.
+            return false;
+        }
+        // We remove the matched word from source to ensure each search word
+        // is matched uniquely.
+
+        // A later search term won't ever uniquely match a term that's already been
+        // matched (and removed) if the original matching search term had another
+        // match. Taking query words in reverse lexicographic order and finding for
+        // each, one by one, the lexicographically last possible prefix match ensures
+        // every query word is matched with the "best" possible source word, given
+        // the context of all other query words.
+        source_words.splice(match_index, 1);
```

Comment:
I still don't feel that this explains *why* this algorithm works correctly. It restates the algorithm, and then basically claims that it *does* work correctly, but doesn't really give reasoning for why.

What you've written down is a certain greedy algorithm: it starts constructing a match of query words to source words, when it has choices it makes them in a certain greedy way, and if it ever finds it can't go on then it throws up its hands and declares there was no solution. To say this algorithm is correct is basically to claim a little theorem: one that says that if this algorithm doesn't successfully construct a match, then there can't have been any possible match at all.

And so the information that should be in a comment somewhere in this function's code is… it doesn't have to be a full detailed proof, not the kind that would necessarily be accepted on a problem set in an algorithms class, but there needs to be a sketch of a proof, or at least a hint at the key ideas in it.

One way to see this doesn't suffice as a proof sketch is to consider: what happens if you go in forward order, rather than backward? So write "Taking query words in lexicographic order" and "the lexicographically first possible prefix match". The resulting explanation would be just as convincing, right? There's nothing this says that would make less sense that way.

Yet if you did that, it turns out the algorithm would not work. So if this reasoning proves anything, then it proves too much — it proves a broken algorithm works, as well as the working algorithm. A valid proof sketch will have something in it that makes use of the fact the algorithm goes specifically in *reverse* lexicographic order.

Reply:
Thanks for the detailed feedback / explanation! I did have a pretty thorough proof written out earlier, but felt it was way too long to have as a comment here, so I ended up with this. Now I've replaced this with an edited down version of the complete proof, which should hopefully work as a proof sketch comment.